docs/fn-2.0/README.md

```diff
@@ -0,0 +1,416 @@
+
+* http: /x -> trigger maps /x -> function f1 with config c1
+* http: /y …
+* schedule: run function f2 every day at 8am
+* queue: listen to topic t1 filtering on ff1, for every event run function f3
+
+All of these things make up the application. Applications also define 
+additional annotations and configuration to be added to each trigger.
+
+## Components
+
+The following components make up the Fn platform. 
+
+* All components can run separately
+* They all have their own main
+* One main main to tie them all into one bin for a nice developer experience.
+
+### Runner
+
+The Runner does nothing more than take an event message, execute the function and return results.
+
+* Main endpoint `/run` with CloudEvent as input.
+* When called, runs a function and returns the results, as a CloudEvent.
+* If busy, returns 503.
+* Event input contains everything it needs to run, including config/secrets.
+  * TODO: Need to be sure we don’t log any secrets
+* Logs can be retrieved separately via `/log` endpoint so main response can be returned faster.
+
+### Function Registry
+
+A service that gives functions an addressable URL, eg:
+
+```
+https://functions.somewhere.com/NAMESPACE/FUNCNAME:VERSION-TAG
+```
+
+* Has metadata required to run a function provided by the creator of the function
+* All functions registered here must accept and return CloudEvent messages
+* This leads to reusable/shareable functions with no dependencies on other components
+* Could potentially run as a separate service like Docker Hub
+
+Q: Seems like a lot to just be a small layer on top of Docker Hub?
+A: It may seem this way now for just a few fields, but down the road, there could many additional things
+  like forcing functions to pass certain tests (eg: I/O formats), defining expected input parameters (can reject 
+  before starting if input is wrong), etc.
+
+### API
+
+CRUD for all the things.
+
+* Namespaces
+* Apps
+* Triggers
+* Functions
+* Calls
+* Logs
+
+### Function Router/LB
+
+Deals with sending event messages to the best runner.
+
+Same main endpoint as a runner, `/run` with CloudEvent input.
+
+### Trigger Manager
+
+The trigger manager is the main entrypoint for an event to enter the Fn platform.
+
+Main endpoint is `/run` with CloudEvent input. 
+
+Flow:
+
+* Input is a CloudEvent.
+* Looks up matching trigger from API based on event.source (maybe other params too).
+* Calls out to LB (or Runner if no LB) to run function.
+
+See **Trigger** above for more information.
+
+### Event Sources
+
+#### HTTP Router
+
+Turns http requests into trigger cloud events (see flow ex. below),
+passes them off to trigger manager / LB / runner.
+
+#### Scheduler
+
+On schedule set by user, the trigger is fired.
+Scheduler would have its own API to manage schedule -> trigger mappings.
+
+#### Queue
+
+Listens on a stream/topic/channel, can have a filter
```

Comment:
For queue what happens with the response if there is any?

Reply:
Comparing to other serverless offerings, queue triggering is completely async (good example here would be an AWS SNS).

Reply:
So is the idea with queues then that any calls have a side affect that is either in logs or in a db somewhere etc?

Reply:
Yes, there is no response.

Reply:
Interesting curious to see how we implement that..

Reply:
Ex: if there is no response how do we know when the container has dealt with the event.

Reply:
I think that internally the runner module would generate a response (it treats all invocations the same) and it's up to the queue event source to either toss out the response or it can do whatever it wants, like, say, put a message onto a different topic with the response or just the call id or post it to imgur. out of the box, we plan to have a 'queue trigger' that would probably act like our async does now, but conceptually it's not quite the same since it would be queueing events to the queue and not through fn (we can add API sugar to the queue source code if we want)